x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SLMGetExpiredSnapshotsAction.java

```diff
@@ -0,0 +1,304 @@
+import org.elasticsearch.action.support.TransportAction;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.repositories.GetSnapshotInfoContext;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.Repository;
+import org.elasticsearch.repositories.RepositoryData;
+import org.elasticsearch.repositories.RepositoryMissingException;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.snapshots.SnapshotState;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.slm.SnapshotLifecyclePolicy;
+import org.elasticsearch.xpack.core.slm.SnapshotRetentionConfiguration;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+public class SLMGetExpiredSnapshotsAction extends ActionType<SLMGetExpiredSnapshotsAction.Response> {
+
+    public static final SLMGetExpiredSnapshotsAction INSTANCE = new SLMGetExpiredSnapshotsAction();
+
+    private static final Logger logger = LogManager.getLogger(SLMGetExpiredSnapshotsAction.class);
+
+    private SLMGetExpiredSnapshotsAction() {
+        super("cluster:admin/slm/execute/get_expired_snapshots", Writeable.Reader.localOnly());
+    }
+
+    public static class LocalAction extends TransportAction<Request, Response> {
+        private final RepositoriesService repositoriesService;
+        private final Executor retentionExecutor;
+        private final ThreadContext threadContext;
+
+        private static final Logger logger = SLMGetExpiredSnapshotsAction.logger;
+
+        @Inject
+        public LocalAction(TransportService transportService, RepositoriesService repositoriesService, ActionFilters actionFilters) {
+            super(INSTANCE.name(), actionFilters, transportService.getTaskManager());
+            this.repositoriesService = repositoriesService;
+            final var threadPool = transportService.getThreadPool();
+            this.retentionExecutor = threadPool.executor(ThreadPool.Names.MANAGEMENT);
+            this.threadContext = threadPool.getThreadContext();
+        }
+
+        private static class ResultsBuilder {
+            private final Map<String, List<Tuple<SnapshotId, String>>> resultsByRepository = ConcurrentCollections.newConcurrentMap();
```

Comment:
nit: we could use the `ConcurrentMap` interface directly here to indicate thread-safety and atomicity guarantees are needed

Reply:
Eh we _could_ but that makes the declaration long enough that it splits over two lines. Moreover we never mention its type again in its 11-line scope.